Common/Orders/Slippage/MarketImpactSlippageModel.cs

```diff
@@ -0,0 +1,221 @@
+using System.Linq;
+using QuantConnect.Interfaces;
+using QuantConnect.Data.Consolidators;
+using QuantConnect.Data.Market;
+using QuantConnect.Securities;
+using QuantConnect.Data;
+using QuantConnect.Indicators;
+using MathNet.Numerics.Statistics;
+
+namespace QuantConnect.Orders.Slippage
+{
+    /// <summary>
+    /// Slippage model that mimic the effect brought by market impact,
+    /// i.e. consume the volume listed in the order book
+    /// </summary>
+    /// <remark>Almgren, R., Thum, C., Hauptmann, E., & Li, H. (2005). 
+    /// Direct estimation of equity market impact. Risk, 18(7), 58-62.</remark>
+    /// <remark>The default parameters are calibrated around 2 decades ago,
+    /// the trading time effect is not accounted (volume near market open/close is larger),
+    /// the market regime is not taken into account,
+    /// and the market environment does not have many market makers at that time,
+    /// so it is recommend to recalibrate with reference to the original paper.</remark>
+    public class MarketImpactSlippageModel : ISlippageModel
+    {
+        private IAlgorithm _algorithm;
+        private readonly double _alpha;
+        private readonly double _beta;
+        private readonly double _gamma;
+        private readonly double _eta;
+        private readonly double _delta;
+        private Dictionary<Symbol, SymbolData> _symbolDataPerSymbol = new();
+        private Random _random = new(50);
+
+        /// <summary>
+        /// Instantiate a new instance of MarketImpactSlippageModel
+        /// </summary>
+        /// <param name="algorithm">IAlgorithm instance</param>
+        /// <param name="alpha">exponent of the permanent impact function</param>
+        /// <param name="beta">exponent of the temporary impact function</param>
+        /// <param name="gamma">coefficient of the permanent impact function</param>
+        /// <param name="eta">coefficient of the temporary impact functio</param>
+        /// <param name="delta">the liquidity scaling factor for permanent impact</param>
+        public MarketImpactSlippageModel(IAlgorithm algorithm, double alpha = 0.891d, double beta = 0.600d,
+                                         double gamma = 0.314d, double eta = 0.142d, double delta = 0.267d)
+        {
+            _algorithm = algorithm;
+            _alpha = alpha;
+            _beta = beta;
+            _gamma = gamma;
+            _eta = eta;
+            _delta = delta;
+        }
+
+        /// <summary>
+        /// Slippage Model. Return a decimal cash slippage approximation on the order.
+        /// </summary>
+        public decimal GetSlippageApproximation(Security asset, Order order)
+        {
+            if (!_symbolDataPerSymbol.TryGetValue(asset.Symbol, out var symbolData))
+            {
+                _symbolDataPerSymbol.Add(asset.Symbol, new SymbolData(_algorithm, asset.Symbol));
+                symbolData = _symbolDataPerSymbol[asset.Symbol];
+            }
+
```

Comment:
`AvgVolume` can be zero (e.g. Forex). We need to check for zero right here, and return 0m.